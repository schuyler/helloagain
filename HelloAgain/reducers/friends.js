'use strict';

import { 
  UPDATE_FRIEND,
  TOGGLE_ACTIVE,
  CONTACTS_LOADED,
  MARK_AS_CONTACTED
} from '../actions/types';

const _id = (friend) => {
  return friend.recordID
}

const findFriend = (state, friend) => {
  return state[_id(friend)] 
}

const queueTail = (state) => {
  const ranks = Object.values(state).map((f) => {
    return f.rank || 0
  })
  return Math.max(0, ...ranks)
}

const defaultValues = (state) => {
  return {
    rank: queueTail(state) + 1,
    isActive: false
  }
}

const markContacted = (state, friend) => {
  return {
    ...friend,
    rank: queueTail(state) + 1,
    contactedAt: Date.now()
  }
}

const updateFriend = (state, update, useDefaults) => {
  let id = _id(update)
  let defaults = useDefaults ? defaultValues(state) : undefined
  let newFriend = {...defaults, ...state[id], ...update}
  return {...state, [id]: newFriend}
}

const importContacts = (state, contacts) => {
  let newState = {...state}
  contacts.forEach((item) => {
    let id = _id(item)
    newState[id] = {...state[id], ...item}
  })
  return newState
}

const friends = (state = {}, action) => {
  switch (action.type) {
    case CONTACTS_LOADED:
      return importContacts(state, action.contacts)
    case TOGGLE_ACTIVE:
      let existingFriend = findFriend(state, action.friend)
      let isActive = (existingFriend ? existingFriend.isActive : false)
      let update = {...action.friend, isActive: !isActive}
      // Set default values (if needed) when toggling a friend
      return updateFriend(state, update, true)
    case MARK_AS_CONTACTED:
      let asContacted = markContacted(state, action.friend)
      return updateFriend(state, asContacted)
    case UPDATE_FRIEND:
      return updateFriend(state, action.friend)
    default:
      return state
  }
}

export default friends
