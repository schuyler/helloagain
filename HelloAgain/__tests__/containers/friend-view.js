'use strict'

import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme'
import configureStore from 'redux-mock-store'

import { copyFriendsFixture } from "../fixtures/friends"
import FriendView from "../../containers/friend-view"

const mockStore = configureStore([])

it('frienders', () => {
  const fixture = copyFriendsFixture()
  Object.values(fixture).forEach((f) => {f.isActive = true})
  const store = mockStore({friends: fixture})
  const friend = Object.values(fixture)[0]
  renderer.create(
    <FriendView store={store} item={friend} />
  )
})

it('marks a friend contacted', () => {
  const fixture = copyFriendsFixture()
  const store = mockStore({friends: fixture})
  const friend = Object.values(fixture)[0]
  const wrapper = shallow(
    <FriendView store={store} item={friend} />
  )
  wrapper.props().onContactedPress(friend)
  const actions = store.getActions()
  expect(actions).toMatchSnapshot()
})
