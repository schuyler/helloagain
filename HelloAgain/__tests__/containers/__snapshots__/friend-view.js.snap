exports[`test marks a friend contacted 1`] = `
Array [
  Object {
    "friend": Object {
      "familyName": "Copernicus",
      "givenName": "Nicolaus",
      "helloAgainID": "nicolaus-copernicus-b594",
      "recordID": 1,
    },
    "type": "MARK_AS_CONTACTED",
  },
]
`;
